iOS_SDK/OneSignalSDK/Source/UNUserNotificationCenter+OneSignal.m

```diff
@@ -170,42 +173,52 @@ - (void)onesignalUserNotificationCenter:(UNUserNotificationCenter *)center
         return;
     }
 
-    [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:@"onesignalUserNotificationCenter:willPresentNotification:withCompletionHandler: Fired!"];
+    [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:[NSString stringWithFormat:@"onesignalUserNotificationCenter:willPresentNotification:withCompletionHandler: Fired! %@", notification.request.content.body]];
     
-    NSDictionary * userInfo = notification.request.content.userInfo;
-    OSNotificationPayload *payload = [OSNotificationPayload parseWithApns:userInfo];
-    NSString *uuid = [payload additionalData][ONESIGNAL_IAM_PREVIEW];
+    [OneSignal handleWillPresentNotificationInForegroundWithPayload:notification.request.content.userInfo withCompletion:^(OSNotificationDisplayType displayType) {
+        finishProcessingNotification(notification, center, displayType, completionHandler, self);
+    }];
+}
 
-    NSUInteger completionHandlerOptions = 0;
-    if (!uuid) {
-        switch (OneSignal.notificationDisplayType) {
-            case OSNotificationDisplayTypeSilent: completionHandlerOptions = 0; break; // Nothing
-            case OSNotificationDisplayTypeNotification: completionHandlerOptions = 7; break; // Badge + Sound + Notification
-            default: break;
-        }
+// To avoid complications caused by swizzling, this is implemented as a C function
```

Comment:
Interesting, can you be more specific on what making a C function avoids with swizzling? No need to research and provide exact details, but if you have an understand it would be good to note that here.

Reply:
We are trying to pass self as "instance" which is the swizzled type of OneSignalUNUserNotificationCenter. This crashes when done in objective-c but works when done in C. I am not much more familiar with it than that (I based this off of the previous implementation of the display type delegate)